core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -267,376 +302,264 @@ protected void doRun() throws Exception {
-                    );
-                }
-            } catch (Throwable t) {
-                // if we failed the mappings anywhere, we need to fail the shards for this index, note, we safeguard
-                // by creating the processing the mappings on the master, or on the node the mapping was introduced on,
-                // so this failure typically means wrong node level configuration or something similar
-                for (IndexShard indexShard : indexService) {
-                    ShardRouting shardRouting = indexShard.routingEntry();
-                    failAndRemoveShard(shardRouting, indexService, true, "failed to update mappings", t);
-                }
+            Set<ShardRouting> shardRoutings = new HashSet<>();
+            for (Shard shard : indexService) {
+                shardRoutings.add(shard.routingEntry());
             }
-        }
-    }
-
-    private boolean processMapping(String index, MapperService mapperService, String mappingType, CompressedXContent mappingSource) throws Throwable {
-        // refresh mapping can happen when the parsing/merging of the mapping from the metadata doesn't result in the same
-        // mapping, in this case, we send to the master to refresh its own version of the mappings (to conform with the
-        // merge version of it, which it does when refreshing the mappings), and warn log it.
-        boolean requiresRefresh = false;
-        try {
-            DocumentMapper existingMapper = mapperService.documentMapper(mappingType);
-
-            if (existingMapper == null || mappingSource.equals(existingMapper.mappingSource()) == false) {
-                String op = existingMapper == null ? "adding" : "updating";
-                if (logger.isDebugEnabled() && mappingSource.compressed().length < 512) {
-                    logger.debug("[{}] {} mapping [{}], source [{}]", index, op, mappingType, mappingSource.string());
-                } else if (logger.isTraceEnabled()) {
-                    logger.trace("[{}] {} mapping [{}], source [{}]", index, op, mappingType, mappingSource.string());
-                } else {
-                    logger.debug("[{}] {} mapping [{}] (source suppressed due to length, use TRACE level if needed)", index, op, mappingType);
-                }
-                mapperService.merge(mappingType, mappingSource, MapperService.MergeReason.MAPPING_RECOVERY, true);
-                if (!mapperService.documentMapper(mappingType).mappingSource().equals(mappingSource)) {
-                    logger.debug("[{}] parsed mapping [{}], and got different sources\noriginal:\n{}\nparsed:\n{}", index, mappingType, mappingSource, mapperService.documentMapper(mappingType).mappingSource());
-                    requiresRefresh = true;
-                }
+            indicesService.removeIndex(indexService.index(), "removing index (mapping update failed)");
+            for (ShardRouting shardRouting : shardRoutings) {
```

Comment:
in the case of createIndices, this should send shard failed for the shard routing that triggered the index creation, but it doesn't because the indexService is empty. I think the interaction with shard failures is too brittle/tricky. My suggestion would be to just return an exception on failure and let the caller deal with it in the right way. 
PS - maybe this signals a testing gap as well..